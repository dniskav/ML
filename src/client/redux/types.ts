import {
  Action,
  PayloadAction,
  TypeConstant,
} from "typesafe-actions";

export interface IReducerAction<TPayload>
  extends Action<TypeConstant>,
  PayloadAction<TypeConstant, TPayload> { }

export interface IApplicationState {
  miscellaneous: {
    itemsLoader: boolean;
    BreadCrumbs: {
      bName: string,
      bUrl: string,
    },
  },
  items: {
    results: IResultItem[],
    currentItem: IResultItem,
  }
}

export interface IResultItem {
  id: string;
  image: { 
    src: string;
    alt: string;
  };
  price: string;
  freeShipping: boolean;
  currencyId: string;
  data: {
    description: string;
    status: string;
  };
  location: string;
};

export interface IResultCollection {
  results: IResultItem[];
};
export interface ICurrentItem {
  condition: string;
  price: string;
  sold_quantity: string;
  title: string;
  pictures: any[];
  currency_id: string;
  description: string;
};
