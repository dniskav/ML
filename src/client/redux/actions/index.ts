import { action } from 'typesafe-actions';
import axios from 'axios';
export const GET_ITEMS_LIST = '@@actions/GET_ITEMS_LIST';
export const SET_ITEMS_LIST = '@@actions/SET_ITEMS_LIST';
export const SET_ITEM = '@@actions/SET_ITEM';
export const SET_DETAIL_ITEM = '@@actions/SET_DETAIL_ITEM';
export const ITEMS_LOADER = '@@actions/ITEMS_LOADER';
export const SET_BREADCRUMBS = '@@actions/SET_BREADCRUMBS';

export const setItem = (item: any) => action(SET_ITEM, item);

export const itemsListLoader = (status: boolean) =>
  action(ITEMS_LOADER, status);

export const getitemsList = (items: any) => action(GET_ITEMS_LIST, items);

export const setitemsList = (items: any) => action(SET_ITEMS_LIST, items);

export const setDetailItem = (item: any) => action(SET_DETAIL_ITEM, item);

export const setBreadcrumbs = (bName: string, bUrl: string) => action(SET_BREADCRUMBS, { bName, bUrl});

export const fetchItemsList = (query: any) => {
  return async (dispatch: any, getState: any) => {
    dispatch(itemsListLoader(true));
    const q = `https://api.mercadolibre.com/sites/MLA/search?q=${query}`;
    try {
      const itemsListRes = await axios.get(q, {});
      dispatch(setitemsList(itemsListRes.data.results));
    } catch (err) {
      console.warn(`ðŸ˜¢ ${err}`, 5);
    } finally {
      dispatch(itemsListLoader(false));
    }
  };
};

export const fetchDetailItem = (id: any) => {
  return async (dispatch: any, getState: any) => {
    dispatch(itemsListLoader(true));
    const q = `https://api.mercadolibre.com/items/${id}`;
    try {
      const detailItemRes = await axios.get(q, {});
      const description = await axios.get(`${q}/description`, {});
      dispatch(setDetailItem({...detailItemRes.data, description: description.data.plain_text}));
      dispatch(setBreadcrumbs(detailItemRes.data.title, `${window.location.origin}/items/${id}`))
    } catch (err) {
      console.warn(`ðŸ˜¢ ${err}`);
    } finally {
      dispatch(itemsListLoader(false));
    }
  };
};

