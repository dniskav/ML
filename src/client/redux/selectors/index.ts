import { createSelector } from 'reselect';
import { IApplicationState } from '../types';

const itemsSelector = (state: IApplicationState) => state.items;
const miscSelector = (state: IApplicationState) => state.miscellaneous;

export const getResults = createSelector(
  itemsSelector,
  (state: any) => state.results,
);

export const getCurrentItem = createSelector(
  itemsSelector,
  (state: any) => state.currentItem,
);

export const getBreadcrumbs = createSelector(
  miscSelector,
  (state: any) => state.breadcrumbs,
);