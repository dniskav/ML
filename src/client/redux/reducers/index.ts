import { combineReducers } from 'redux';
import { IApplicationState } from '../types';
import itemsReducer from './ItemsReducer';
import { miscReducer } from './miscReducer';

const rootReducer = combineReducers<IApplicationState>({
  items: itemsReducer,
  miscellaneous: miscReducer,
});

export default rootReducer;
