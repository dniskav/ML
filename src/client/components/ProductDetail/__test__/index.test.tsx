import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import PDP from '../index';
import { ICurrentItem } from '../../../redux/types';

const currentItem: ICurrentItem = {
  condition: "new",
  currency_id: "",
  description: "",
  pictures: [],
  price: "",
  sold_quantity: "",
  title: ""
};

jest.mock('react-redux', () => {
  return {
    useDispatch: () => jest.fn(),
    useSelector: (selectorName: any) => {
          return currentItem;
      }
  };
});

jest.mock('react-router', () => ({
  useParams: () => jest.fn({ id: '1' } as any),
}));

describe('<PDP />', () => {
  it("render success", () => {
    const wrapper: ShallowWrapper = shallow(<PDP />);
    expect(wrapper).toMatchSnapshot();
  });
});
