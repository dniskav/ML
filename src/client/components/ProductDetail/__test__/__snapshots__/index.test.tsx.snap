// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PDP /> render success 1`] = `ShallowWrapper {}`;
